### PR TITLE
Consistent symbols

### PR DESCRIPTION
**Pull request by alexnick83**
This PR attempts to address the fact that symbols are being converted to string and back to symbols again in the whole DaCe stack. This is a serious issue because important symbolic assumptions are lost. A characteristic example is the NPBench benchmark `stockham_fft` which has symbolic sizes with either powers or logarithms. This sample can only work with the array sizes being explicitly defined as positive integers and the loop-indices implicitly defined as non-negative integers.

### CODE REVIEW

File: dace/sdfg/sdfg.py [hunk 1]
```diff
@@ -157,8 +157,29 @@ def replace(self, name: str, new_name: str, replace_keys=True) -> None:
...
+            if new_idx + length < len(condition):
+                if re.match("[A-Za-z0-9_]+$", condition[new_idx + length]):
+                    matched = False
+            last_idx, new_idx = new_idx, condition.find(name, new_idx + length)
+            if matched:
+                new_condition += new_name
+                last_idx += length
+        if last_idx < len(condition):
+            new_condition += condition[last_idx:]
+        self.condition.as_string = new_condition
```
Comment (tbennun) [inline dace/sdfg/sdfg.py:182]: shouldn't we just convert the condition to AST (we do this just above for assignments) and then use ASTFindAndReplace?

File: dace/symbolic.py [hunk 2]
```diff
@@ -854,3 +854,43 @@ def persistent_load(self, pid):
...
+
+def inequal_symbols(a: Union[sympy.Expr, Any],
+                    b: Union[sympy.Expr, Any]) -> bool:
+    """ Compares 2 symbolic expressions and returns True if they are not equal.
+    """
+    if not isinstance(a, sympy.Expr) or not isinstance(b, sympy.Expr):
+        return a != b
+    else:
+        a, b = equalize_symbols(a, b)
+        return (a - b).simplify() != 0
```
Comment (tbennun) [inline dace/symbolic.py:896]: do we need the simplify here? why not ask `a != b`? please leave a comment in the code too (this is one of the things that can be easily forgotten).

File: dace/symbolic.py [hunk 3]
```diff
@@ -854,3 +854,43 @@ def persistent_load(self, pid):
...
+        repldict = dict()
+        for name in common_names:
+            repldict[b_syms[name]] = a_syms[name]
+        b = b.subs(repldict)
+    return a, b
+
+
+def inequal_symbols(a: Union[sympy.Expr, Any],
+                    b: Union[sympy.Expr, Any]) -> bool:
+    """ Compares 2 symbolic expressions and returns True if they are not equal.
```
Comment (tbennun) [inline dace/symbolic.py:890]:
```suggestion
    """
    Compares 2 symbolic expressions and returns True if they are not equal.
```

File: dace/transformation/dataflow/redundant_array.py [hunk 4]
```diff
@@ -49,10 +49,15 @@ def _validate_subsets(edge: graph.MultiConnectorEdge,
...
+                src_expr = src_subset.num_elements()
+                src_expr_exact = src_subset.num_elements_exact()
+                dst_expr = dst_subset.num_elements()
+                dst_expr_exact = dst_subset.num_elements_exact()
+                if (src_expr != dst_expr and symbolic.inequal_symbols(
+                        src_expr_exact, dst_expr_exact)):
                     raise ValueError(
-                        "Source subset is missing (dst_subset: {}, "
-                        "src_shape: {}".format(dst_subset, desc.shape))
+                        "Destination subset is missing (src_subset: {}, "
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py:59]: what happened here with the exception?

File: dace/transformation/dataflow/redundant_array.py [hunk 5]
```diff
@@ -187,15 +197,59 @@ def can_be_applied(graph, candidate, expr_index, sdfg, strict=False):
...
             # In strict mode, make sure the memlet covers the removed array
-            edge = graph.edges_between(in_array, out_array)[0]
-            subset = copy.deepcopy(edge.data.subset)
+            subset = copy.deepcopy(e1.data.subset)
             subset.squeeze()
             shape = [sz for sz in in_desc.shape if sz != 1]
             if any(m != a for m, a in zip(subset.size(), shape)):
                 return False
+
+            # NOTE: Library node check
```
Comment (tbennun) [inline dace/transformation/dataflow/redundant_array.py:213]: with this code, can we remove the reduce check below?

**Review by tbennun — CHANGES_REQUESTED** [review]

File: dace/frontend/python/memlet_parser.py [hunk 6]
```diff
@@ -85,14 +92,21 @@ def _fill_missing_slices(das, ast_ndslice, array, indices):
...
-            if (rb < 0) == True:
-                rb += array.shape[indices[idx]]
-            if (re < 0) == True:
-                re += array.shape[indices[idx]]
+            rs = _parse_dim_atom(das, dim[2] or 1)
+            # NOTE: try/except for cases where rb/re are not symbols/numbers
+            try:
+                if (rb < 0) == True:
+                    rb += array.shape[indices[idx]]
+            except:
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/memlet_parser.py:103]: I think it is ValueError?

[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/memlet_parser.py:103]: No +/- operation defined for 'Data' and 'int', something like this was the issue

File: dace/frontend/python/newast.py [hunk 7]
```diff
@@ -798,7 +809,7 @@ def _add_access(
             sym_rng = []
             for i, r in enumerate(rng):
                 for s, sr in self.symbols.items():
-                    if s in symbolic.symlist(r).keys():
+                    if s in symbolic.symlist(r).values():
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/newast.py:812]: Let me clarify. Originally, `ProgramVisitor.symbols` was `Dict[str, Range]`, where the key was the symbol name, similar to `SDFG.symbols`. To make symbols consistent, this was changed to `Dict[Symbol, Range]`. Now the key is the Symbol object itself. `symbolic.symlist(Range)` return `Dict[str, Symbol]`, where the key is the symbol name and the value is the Symbol object. In the above snipper, `s` is now the Symbol object, therefore, I have to look into the values of the dictionary.

File: dace/symbolic.py (hunk 3 above)
Comment (alexnick83) [reply dace/symbolic.py:890]: I don't understand the suggestion? Is it to move the text one line below? Why?

File: dace/symbolic.py (hunk 2 above)
Comment (alexnick83) [reply dace/symbolic.py:896]: From SymPy's documentation (https://docs.sympy.org/latest/tutorial/gotchas.html):
```python
a = (x + 1)**2
b = x**2 + 2*x + 1
simplify(a - b)
0
```

Comment (alexnick83) [reply dace/symbolic.py:896]: And yes we need the `simplify` to try to remove `ceiling`, `floor`, etc. when the assumptions allow it

File: dace/transformation/dataflow/redundant_array.py (hunk 4 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py:59]: If you mean about the message, it is obviously a copy-paste issue. It is exactly the same code as below, but the message should be different.

File: dace/transformation/dataflow/redundant_array.py (hunk 5 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py:213]: The reduce check is for a different issue. It is related to reductions with the `keepdims` keyword enabled. To generate the correct SDFG, we first generate the reduction ignoring `keepdims` and then we copy to the final output. We are not checking any relation between the input and the output. Also the new library node check does not ensure that the "intermediate" reduction output in the case of `keepdims=True` is not removed.

File: dace/symbolic.py (hunk 2 above)
Comment (alexnick83) [reply dace/symbolic.py:896]: (`stockham_fft` is triggering all these issues)

File: dace/frontend/python/memlet_parser.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tbennun) [reply dace/frontend/python/memlet_parser.py:103]: yes, I mean that this is an exception that is too wide (takes things e.g., KeyboardInterrupt as well). this is disallowed in our coding guidelines. I was proposing to replace it with `except (TypeError, ValueError):`.

File: dace/transformation/dataflow/redundant_array.py (hunk 5 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py:213]: Let me also clarify that the problem with `keepdims=True` is that our Reduce node expansions are not compatible with writing directly to the final keepdims-enabled output. And we have to many reduce expansions to fix them now.

File: dace/symbolic.py (hunk 2 above)
Comment (tbennun) [reply dace/symbolic.py:896]: ok, thanks. please add a short comment that it simplifies the ceiling expressions etc.

File: dace/symbolic.py (hunk 3 above)
Comment (tbennun) [reply dace/symbolic.py:890]: just a matter of consistent style.
```
"""
doc
:param x:
:param y:
:return:
:note:
:raises z:
"""
```

File: dace/transformation/dataflow/redundant_array.py (hunk 4 above)
Comment (tbennun) [reply dace/transformation/dataflow/redundant_array.py:59]: yes

File: dace/transformation/dataflow/redundant_array.py (hunk 5 above)
Comment (tbennun) [reply dace/transformation/dataflow/redundant_array.py:213]: ok

File: dace/sdfg/sdfg.py (hunk 1 above)
Comment (alexnick83) [reply dace/sdfg/sdfg.py:182]: I will give it a shot

Comment (alexnick83) [reply dace/sdfg/sdfg.py:182]: Done

File: dace/symbolic.py (hunk 2 above)
Comment (alexnick83) [reply dace/symbolic.py:896]: Done

File: dace/symbolic.py (hunk 3 above)
Comment (alexnick83) [reply dace/symbolic.py:890]: Done

File: dace/transformation/dataflow/redundant_array.py (hunk 4 above)
Comment (alexnick83) [reply dace/transformation/dataflow/redundant_array.py:59]: Done

File: dace/frontend/python/memlet_parser.py [hunk 8]
```diff
@@ -85,14 +92,21 @@ def _fill_missing_slices(das, ast_ndslice, array, indices):
...
+                pass
+            try:
+                if (re < 0) == True:
+                    re += array.shape[indices[idx]]
+            except:
```
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/memlet_parser.py:108]: Done

File: dace/frontend/python/memlet_parser.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexnick83) [reply dace/frontend/python/memlet_parser.py:103]: Done

**Review by tbennun — APPROVED** [review]
LGTM

**Pull request merged**